src/lib/matchers.rs

```diff
@@ -335,13 +336,26 @@ impl<T: Clone + Debug + Eq + 'static> MappingStore<T> {
         // Combined update, insert, align and move phases.
         let tmp_to_arena = self.to_arena.clone();
         for x in root.breadth_first_traversal(&tmp_to_arena) {
-            if x.is_root(&self.to_arena) {
-                continue;
-            }
-            let y = self.to_arena[x].parent().unwrap(); // x is not a root.
             let mut w = root; // Overwritten later.
             // Insertion phase.
-            if !new_mappings.to.contains_key(&x) {
+            if !new_mappings.to.contains_key(&x) && x.is_root(&self.to_arena) {
+                let k = self.find_pos(x, &new_mappings, &to_in_order);
+                debug!("Edit script: INS {} {:?} No parent",
+                       self.to_arena[x].label,
+                       self.to_arena[x].value);
+                w = self.from_arena.new_node(self.to_arena[x].value.clone(),
```

Comment:
What exactly are we cloning here? [I'm not saying it's wrong, but I'm not really sure if we need to clone, or whether a reference might do.]

Reply:
`value` is of a type `T`, it's dependent on the grammar...

Reply:
OK.